src/merge_operator.rs

```diff
@@ -0,0 +1,373 @@
+///         let increment = u64::from_le_bytes(value.as_ref().try_into().unwrap());
+///         Ok(Bytes::copy_from_slice(&(existing + increment).to_le_bytes()))
+///     }
+/// }
+/// ```
+pub trait MergeOperator {
+    /// Merges the existing value with a new value to produce a combined result.
+    ///
+    /// This method is called during reads and compactions to combine multiple merge operands
+    /// into a single value. The implementation must be associative to ensure correct behavior.
+    ///
+    /// # Arguments
+    /// * `existing_value` - The current accumulated value
+    /// * `value` - The new value to merge with the existing value
+    ///
+    /// # Returns
+    /// * `Ok(Bytes)` - The merged result as bytes
+    /// * `Err(MergeOperatorError)` - If the merge operation fails
+    fn merge(&self, existing_value: Bytes, value: Bytes) -> Result<Bytes, MergeOperatorError>;
+}
+
+pub(crate) type MergeOperatorType = Arc<dyn MergeOperator + Send + Sync>;
+
+pub(crate) struct MergeOperatorIterator<T: KeyValueIterator> {
+    merge_operator: MergeOperatorType,
+    delegate: T,
+    /// Entry from the delegate that we've peeked ahead and buffered.
+    buffered_entry: Option<RowEntry>,
+    /// Whether to merge entries with different expire timestamps.
+    merge_different_expire_ts: bool,
+}
+
+#[allow(unused)]
+impl<T: KeyValueIterator> MergeOperatorIterator<T> {
+    pub(crate) fn new(
+        merge_operator: MergeOperatorType,
+        delegate: T,
+        merge_different_expire_ts: bool,
+    ) -> Self {
+        Self {
+            merge_operator,
+            delegate,
+            buffered_entry: None,
+            merge_different_expire_ts,
+        }
+    }
+}
+
+impl<T: KeyValueIterator> MergeOperatorIterator<T> {
+    async fn merge_with_older_entries(
+        &mut self,
+        key: Bytes,
+        seq: u64,
+        value: Bytes,
+        create_ts: Option<i64>,
+        expire_ts: Option<i64>,
+    ) -> Result<Option<RowEntry>, SlateDBError> {
+        let key = key;
+        let mut merged_value = value;
+        let mut max_create_ts = create_ts;
+        let mut min_expire_ts = expire_ts;
+
+        // Keep looking ahead and merging as long as we find mergeable entries
+        loop {
+            let next = self.delegate.next_entry().await?;
+            match next {
+                Some(next_entry)
+                    if key == next_entry.key
+                        && (self.merge_different_expire_ts
+                            || expire_ts == next_entry.expire_ts) =>
+                {
+                    // Accumulate timestamps
+                    max_create_ts = merge_options(max_create_ts, next_entry.create_ts, i64::max);
+                    min_expire_ts = merge_options(min_expire_ts, next_entry.expire_ts, i64::min);
+                    // For sequence number, there is nothing to accumulate. Just ensure it keeps decreasing
+                    if seq < next_entry.seq {
```

Comment:
do we want to update `seq`?

Reply:
No, since we want the highest one = the first one we've seen

Reply:
I've expanded the note.